test/test_public_bindings.py

```diff
@@ -4,7 +4,331 @@
+            'torch.distributions.Dirichlet', 'torch.utils.data.datapipes.dataframe.DFIterDataPipe',
+            'torch.nn.modules.Softmax', 'torch.distributions.ComposeTransform', 'torch.nn.modules.CrossMapLRN2d',
+            'torch.nn.modules.LazyConvTranspose2d', 'torch.nn.modules.TripletMarginLoss', 'torch.special.logit',
+            'torch.special.expit', 'torch.nn.modules.CosineSimilarity', 'torch.special.erfc',
+            'torch.utils.data.datapipes.iter.Grouper', 'torch.nn.parallel.gather', 'torch.nn.modules.ReflectionPad1d',
+            'torch.special.exp2', 'torch.special.i0e', 'torch.nn.modules.LazyInstanceNorm1d',
+            'torch.distributions.HalfCauchy',
+        ])
+
+        # No new entry should EVER be added to this set, only removed!
+        missing_obj_attr_allowlist = set(['torch.utils.data.datapipes.iter.DFIterDataPipe'])
+
+        def is_not_internal(modname):
+            split_name = modname.split(".")
+            for name in split_name:
+                if name[0] == "_":
+                    return False
+            return True
+
+        # Allow this script to run when
+        #  - Built with USE_DISTRIBUTED=0
+        #  - TensorRT is not installed
+        #  - Until torch.utils.ffi module is removed
+        def cannot_be_skipped(modname):
+            if "distributed" in modname and not torch.distributed.is_available():
+                return False
+            if "fx2trt" in modname and not hasattr(trt, "__version__"):
+                return False
+            if modname == "torch.utils.ffi":
+                return False
+            return True
+
+        missing_all_attr = set()
+        missing_module_attr = set()
+        missing_obj_attr = set()
+
+        def error_handler(bad_pkg):
+            if is_not_internal(bad_pkg) and cannot_be_skipped(bad_pkg):
+                raise RuntimeError(f"Failed to import public package {bad_pkg}")
+
+        for _, modname, ispkg in pkgutil.walk_packages(path=torch.__path__,
+                                                       prefix=torch.__name__ + '.',
+                                                       onerror=error_handler):
+            if ispkg and is_not_internal(modname):
+
+                try:
+                    mod = importlib.import_module(modname)
+                except Exception as e:
+                    if cannot_be_skipped(modname):
+                        raise e
+
+                if not hasattr(mod, "__all__"):
+                    missing_all_attr.add(modname)
+                    continue
+                for el_name in mod.__all__:
+                    if not hasattr(mod, el_name):
+                        missing_obj_attr.add(f"{modname}.{el_name}")
+                        continue
+                    obj = getattr(mod, el_name)
+                    if isinstance(obj, ModuleType):
+                        continue
+                    if not hasattr(obj, "__module__") or obj.__module__ != modname:
```

Comment:
The second clause seems incorrect to me. `__module__` is supposed to point to where the object was defined I believe, rather than where it is exposed. I'm not 100% sure, because this attribute is used very rarely, but that's how it works by default and is what this explanation indicates:
- https://stackoverflow.com/questions/10113892/semantics-of-module
- https://www.python.org/dev/peps/pep-3130/#rationale-for-module

Reply:
Or maybe this is a conscious choice, given the error message ~20 lines below?

Reply:
The reason for this semantic is that:
- The coverage test we use will only consider the functions from the current module: https://github.com/sphinx-doc/sphinx/blob/1f91a0f8bd33eeffe082e31300f94df2ee10454f/sphinx/ext/coverage.py#L166-L176 so we want the `__module__` to be set properly.
- The other thing is that this coverage test is only considering the modules that are not private (starting with an `_`) that are in the website. That means that if we use ` _foo` to define some function that we then expose at the top level (like the c++ code), this won't be covered by the coverage tests.
- For things like https://github.com/pytorch/pytorch/tree/master/torch/nn/modules this is more open ended but we still end up exposing all the modules as part of the public API for for the top level `modules` package and each file in it.

I would be curious if you have ideas of how to improve that?
Maybe if there is a mismatch but the defined `__module__` is another of pytorch's public module, then it is ok (as we know we will get the coverage from there)?

Reply:
By the way, I got some of the inspiration from https://github.com/numpy/numpy/blob/d5f6618d74a427d0ea337a5d1f17ecc0ab7ce8da/numpy/tests/test_public_api.py

Reply:
> The coverage test we use will only consider the functions from the current module

Ah okay, so this is matching Sphinx. I think in NumPy/SciPy that doesn't bother us, because we set (if needed) `.. currentmodule::` in the function listings & autosummary reST files. And we use a separate checker to check for docs completeness which does not rely on Sphinx: [numpy-refguide-check](https://github.com/numpy/numpy/blob/main/tools/refguide_check.py). The more reusable form of that is https://github.com/astropy/pytest-doctestplus

That said, I don't think `__module__` is used much, so I'm fine with keeping it as you have it in your PR now.

Reply:
> Ah okay, so this is matching Sphinx

Of the coverage tool of sphinx in particular. We also use currentmodule for the doc itself, so this is not a problem for the doc/website. Only the coverage part.

> does not rely on Sphinx: numpy-refguide-check. 

We used to have something like that in torch as well and moved to the sphinx one. Do you have any feedback on your feeling about which one is best?
Note that the sphinx coverage tool is just a couple hundred lines of code, so we could have a much more custom coverage tool based on sphinx (to avoid having to do the rst parsing ourselves). Do you think that could work?

Also is there a high level description of what this checker does and how does it make sure that all the non-private functions are considered while ignoring imports?

Reply:
> We used to have something like that in torch as well and moved to the sphinx one. Do you have any feedback on your feeling about which one is best?

I'm not sure, I don't have experience with the Sphinx coverage checker. It's pretty much undocumented, and my experience with Sphinx is that it's hard to get fixes merged, so I'd personally prefer the AstroPy one or a custom one.

> Note that the sphinx coverage tool is just a couple hundred lines of code, so we could have a much more custom coverage tool based on sphinx (to avoid having to do the rst parsing ourselves). Do you think that could work?

Probably - but if you're happy with it today, don't touch it until it breaks?

> Also is there a high level description of what this checker does and how does it make sure that all the non-private functions are considered while ignoring imports?

The numpy one? No separate docs, but the module/function docstrings and code comments are pretty extensive: https://github.com/numpy/numpy/blob/main/tools/refguide_check.py. Look at `find_names` and `get_all_dict` in that file I suggest (docs and `__all__` are compared with those), and then there's a third check for "missing in `__all__`" [here](https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L420). In my experience it's fairly robust.